<a href="https://colab.research.google.com/github/Anastasya2701/GOIT-KATRUSHENKO/blob/main/HW7_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
!pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 4.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-linux_x86_64.whl size=2357275 sha256=d639060aeed4757302cb8c3ad4e36be1ff8e1b0797b726ccab309e1772f88648
  Stored in directory: /root/.cache/pip/wheels/4b/3f/df/6acbf0a40397d9bf3ff97f582cc22fb9ce66adde75bc71fd54
Successfully built scikit-surprise


In [7]:
from surprise import SVD
from surprise import Dataset, Reader
from surprise.model_selection import cross_validate
from surprise.model_selection import GridSearchCV

# Load the movielens-100k dataset (download it if needed).
data = Dataset.load_builtin('ml-100k')

# Use the famous SVD algorithm.
algo = SVD()

cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9276  0.9310  0.9453  0.9384  0.9413  0.9367  0.0065  
MAE (testset)     0.7312  0.7333  0.7430  0.7428  0.7423  0.7385  0.0052  
Fit time          1.38    1.89    1.68    1.41    1.41    1.55    0.20    
Test time         0.12    0.20    0.14    0.13    0.35    0.19    0.08    


{'test_rmse': array([0.9276284 , 0.93096468, 0.94533025, 0.93837089, 0.94129764]),
 'test_mae': array([0.7311792 , 0.73328868, 0.74296324, 0.74276271, 0.7422906 ]),
 'fit_time': (1.3803317546844482,
  1.890488624572754,
  1.6761775016784668,
  1.4055545330047607,
  1.4114482402801514),
 'test_time': (0.1233823299407959,
  0.20369362831115723,
  0.13599276542663574,
  0.12568235397338867,
  0.34581875801086426)}

In [10]:
param_grid = {
    'n_factors': [20, 50, 100],
    'n_epochs': [20, 30],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

gs = GridSearchCV(SVD, param_grid, measures=['rmse', 'mae'], cv=5)

gs.fit(data)

gs.best_score['rmse'], gs.best_params['rmse']

(0.9275778866521411,
 {'n_factors': 100, 'n_epochs': 30, 'lr_all': 0.005, 'reg_all': 0.1})

In [11]:
from surprise import SVDpp, NMF

In [12]:
algo_svdpp = SVDpp()
cross_validate(algo_svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)


Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9144  0.9167  0.9275  0.9218  0.9242  0.9209  0.0048  
MAE (testset)     0.7179  0.7212  0.7281  0.7229  0.7239  0.7228  0.0033  
Fit time          29.33   28.30   28.85   27.61   27.70   28.36   0.66    
Test time         6.81    4.69    4.41    5.55    4.38    5.17    0.92    


{'test_rmse': array([0.91444811, 0.91669845, 0.92749364, 0.92179752, 0.92423587]),
 'test_mae': array([0.717885  , 0.72116961, 0.72806552, 0.72293166, 0.72393878]),
 'fit_time': (29.332239627838135,
  28.304056406021118,
  28.8479163646698,
  27.61334252357483,
  27.697771549224854),
 'test_time': (6.814326763153076,
  4.6905481815338135,
  4.411224126815796,
  5.5532307624816895,
  4.382455825805664)}

In [13]:
algo_nmf = NMF()
cross_validate(algo_nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9685  0.9618  0.9645  0.9623  0.9652  0.9644  0.0024  
MAE (testset)     0.7611  0.7563  0.7581  0.7548  0.7585  0.7577  0.0021  
Fit time          2.22    2.80    2.13    2.14    2.19    2.29    0.25    
Test time         0.17    0.11    0.29    0.12    0.11    0.16    0.07    


{'test_rmse': array([0.96847879, 0.96175608, 0.96448469, 0.96230699, 0.96517893]),
 'test_mae': array([0.76106993, 0.75627724, 0.75814735, 0.75477919, 0.75846891]),
 'fit_time': (2.21737003326416,
  2.7994489669799805,
  2.1322567462921143,
  2.1373372077941895,
  2.1862080097198486),
 'test_time': (0.17091131210327148,
  0.10803413391113281,
  0.2920112609863281,
  0.11934065818786621,
  0.10752153396606445)}

Висновок:

 Побудувавши декілька моделей рекомендаційних систем на основі матричної факторизації, використовуючи бібліотеку Surprise та датасет MovieLens. Після експериментів із різними алгоритмами, такими як SVD, SVD++ і NMF, та виконання крос-валідації отримали наступні результати:

 1. Алгоритм SVD дав хороший баланс між точністю та швидкістю роботи. Під час крос-валідації середнє значення RMSE склало X (0.9276284). Підбір гіперпараметрів за допомогою GridSearchCV дозволив зменшити помилку до Y (0.9275778866521411).
 2. Алгоритм SVD++ показав трохи кращі результати в порівнянні з SVD, але є більш ресурсоємним. Це варто враховувати при роботі з більшими наборами даних або у випадку обмеженого часу на тренування.
 3. Алгоритм NMF виявився менш ефективним у порівнянні з SVD та SVD++, але може бути корисним у випадках, коли потрібна інтерпретація компонент.

Отже, алгоритм SVD із налаштованими гіперпараметрами став оптимальним вибором для цього завдання, оскільки забезпечує хороший компроміс між точністю прогнозів і швидкістю роботи.